File: client/src/types/Patient.ts
type Patient = {
    info: Info;
    visit: Visit[];
    thymus: Thymus[];
    bloodTest: BloodTest[];
    QOL: QOL[];
    QMG: QMG[];
    MG: MG[];
    ADL: ADL[];
    EMG: EMG[];
};

type tablePatient = {
    info: Info;
    visit: Visit;
    thymus: Thymus;
    bloodTest: BloodTest;
    QOL: QOL;
    QMG: QMG;
    MG: MG;
    ADL: ADL;
    EMG: noImageType;
};

// 基本資料
type Info = {
    "ID#": string;
    name: string;
    DOB: string;
    sex: string;
    height: number;
    weight: number;
    other: string;
    attackDate: string;
    beginSymptom: string;
    otherHospitalRecord: {
        recentlyDate: string;
        totalTimes: number;
    };
    otherDisease: string[];
    otherMedicine: string[];
};

// 看診紀錄
type Visit = {
    testDate: string;
    treat: number; // 0 ~ 4
    SBP: number;
    DBP: number;
    selfAssessment: number; // 0 ~ 2
    note: string;
    prescription: {
        // 0 ~ 9
        pyridostigmine: number;
        compesolone: number;
        cellcept: number;
        imuran: number;
        prograf: number;
    };
    examination: {
        // 0 or 1 except MGFAclassification
        ptosis: number;
        diplopia: number;
        dysphagia: number;
        dysarthria: number;
        dyspnea: number;
        limpWeakness: number;
    };
    MGFAclassification: string; // I, II, IIA, IIB, III, IIIA, IIIB, IV, IVA, IVB, V
    status: {
        // 候診狀態
        isWaiting: boolean; // 是否候診中
        description: number; // 候診狀態描述
    };
};

// 胸腺掃描
type Thymus = {
    testDate: string;
    thymusStatus: number; // 0 ~ 3
    thymusDescription: string;
};

// 血液檢查
type BloodTest = {
    testDate: string;
    ACHR: number;
    TSH: number;
    freeThyroxine: number;
    ANA: number;
    uricAcid: number;
};

// 0 ~ 2 except testDate, sum
type QOL = {
    testDate: string;
    frustration: number;
    eyeUsing: number;
    eating: number;
    social: number;
    entertainment: number;
    fullfillFamilyNeeds: number;
    plansNecessity: number;
    jobState: number;
    speaking: number;
    driving: number;
    depression: number;
    walking: number;
    beingInPublicPlaces: number;
    overwhelm: number;
    freshenUp: number;
    sum: number;
};

// 0 ~ 3 except testDate, sum
type QMG = {
    testDate: string;
    doubleVision: number;
    ptosis: number;
    facialMuscle: number;
    swallowing: number;
    speakFluency: number;
    rightArmHeight: number;
    leftArmHeight: number;
    vitalCapacity: number;
    rightHandGrid: number;
    leftHandGrid: number;
    headLift: number;
    rightLegHeight: number;
    leftLegHeight: number;
    sum: number;
};

// 0 ~ 3 except testDate, sum
type MG = {
    testDate: string;
    ptosis: number;
    doubleVision: number;
    eyeClosure: number;
    talking: number;
    chewing: number;
    swallowing: number;
    breathing: number;
    neckFlexion: number;
    shoulderAbduction: number;
    hipFlexion: number;
    sum: number;
};

// 0 ~ 3 except testDate, sum
type ADL = {
    testDate: string;
    talking: number;
    chewing: number;
    swallowing: number;
    breathing: number;
    brushTeethOrCombHair: number;
    ariseFromChair: number;
    doubleVision: number;
    eyelid: number;
    sum: number;
};

type EMG = {
    testDate: string;
    img: Blob;
    nasalis: {
        preActivation: number;
        postActivation: number[];
    };
    abd: {
        preActivation: number;
        postActivation: number[];
    };
    trapezius: {
        preActivation: number;
        postActivation: number[];
    };
};

type noImageType = {
    testDate: string;
    nasalis: {
        preActivation: number;
        postActivation: number[];
    };
    abd: {
        preActivation: number;
        postActivation: number[];
    };
    trapezius: {
        preActivation: number;
        postActivation: number[];
    };
};

type PatientInList = {
    _id: string;
    doctorId?: string;
    nurseId?: string;
    isChecked: boolean;
    info: Info;
};

export type {
    Patient,
    Info,
    Visit,
    Thymus,
    BloodTest,
    QOL,
    QMG,
    MG,
    ADL,
    EMG,
    noImageType,
    tablePatient,
    PatientInList,
};

File: server/models/patient.py
from pydantic import BaseModel, Field
from typing import Optional
from pydantic import BaseModel


class OtherHospitalRecord(BaseModel):
    recentlyDate: str
    totalTimes: int


# 基本資料
class Info(BaseModel):
    ID: str = Field(alias="ID#")
    name: str
    DOB: str
    sex: str
    height: float
    weight: float
    other: str
    attackDate: str
    beginSymptom: str
    otherHospitalRecord: OtherHospitalRecord
    otherDisease: list[str]
    otherMedicine: list[str]


class Prescription(BaseModel):
    pyridostigmine: int
    compesolone: int
    cellcept: int
    imuran: int
    prograf: int


class Examination(BaseModel):
    ptosis: int
    diplopia: int
    dysphagia: int
    dysarthria: int
    dyspnea: int
    limpWeakness: int


class Status(BaseModel):
    isWaiting: bool = Field(default=False)
    description: int = Field(default=0)


# 看診紀錄
class Visit(BaseModel):
    testDate: str
    treat: int
    SBP: float
    DBP: float
    selfAssessment: int
    MGFAclassification: str
    note: str
    prescription: Prescription
    examination: Examination
    status: Status


# 胸腺掃描
class Thymus(BaseModel):
    testDate: str
    thymusStatus: int
    thymusDescription: str


# 血液檢查
class BloodTest(BaseModel):
    testDate: str
    ACHR: float
    TSH: float
    freeThyroxine: float
    ANA: float
    uricAcid: float


# QOL
class QOL(BaseModel):
    testDate: str
    frustration: int
    eyeUsing: int
    eating: int
    social: int
    entertainment: int
    fullfillFamilyNeeds: int
    plansNecessity: int
    jobState: int
    speaking: int
    driving: int
    depression: int
    walking: int
    beingInPublicPlaces: int
    overwhelm: int
    freshenUp: int
    sum: int


# QMG
class QMG(BaseModel):
    testDate: str
    doubleVision: int
    ptosis: int
    facialMuscle: int
    swallowing: int
    speakFluency: int
    rightArmHeight: int
    leftArmHeight: int
    vitalCapacity: int
    rightHandGrid: int
    leftHandGrid: int
    headLift: int
    rightLegHeight: int
    leftLegHeight: int
    sum: int


# MG
class MG(BaseModel):
    testDate: str
    ptosis: int
    doubleVision: int
    eyeClosure: int
    talking: int
    chewing: int
    swallowing: int
    breathing: int
    neckFlexion: int
    shoulderAbduction: int
    hipFlexion: int
    sum: int


# ADL
class ADL(BaseModel):
    testDate: str
    talking: int
    chewing: int
    swallowing: int
    breathing: int
    brushTeethOrCombHair: int
    ariseFromChair: int
    doubleVision: int
    eyelid: int
    sum: int


class Nasalis(BaseModel):
    preactivation: float
    postactivation: list[float]


class Abd(BaseModel):
    preactivation: float
    postactivation: list[float]


class Trapezius(BaseModel):
    preactivation: float
    postactivation: list[float]


# 電生理訊號
class EMG(BaseModel):
    testDate: str
    image: Optional[bytes]
    nasalis: Nasalis
    abd: Abd
    trapezius: Trapezius


# 病患資料
class Patient(BaseModel):
    info: Info
    visit: list[Visit] = Field(default=[])
    thymus: list[Thymus] = Field(default=[])
    bloodTest: list[BloodTest] = Field(default=[])
    qol: list[QOL] = Field(default=[], alias="QOL")
    qmg: list[QMG] = Field(default=[], alias="QMG")
    mg: list[MG] = Field(default=[], alias="MG")
    adl: list[ADL] = Field(default=[], alias="ADL")
    emg: list[EMG] = Field(default=[], alias="EMG")


class PatientForPredict(BaseModel):
    info: Info
    visit: Visit
    thymus: Thymus
    bloodTest: BloodTest
    qol: QOL = Field(alias="QOL")
    qmg: QMG = Field(alias="QMG")
    mg: MG = Field(alias="MG")
    adl: ADL = Field(alias="ADL")
